### PR TITLE
Update Pull Request workflow

### PR DESCRIPTION
**Pull request by calebmshafer**
> This description is work-in-progress and may move to GitHub Issue for better tracking...

This PR contains a number of changes required to update our pull request workflow.

**Status of changes:**
- [ ] Implement 3 hour timeout policy
- [ ] Add easy way for maintainers to queue builds
  - Required for both issue 1 and 2 below
- [ ] Document PR process for both outside contributors, maintainers and writers.
- [ ] Only required iModel.js Docs build on purely "docs/" changes 

**Reason for the change:**

We currently have 2 major issues with our PR process:

1. We run too many unnecessary builds in order to complete a single pull request
2. Fork-based PRs are annoying to get completed due to the basically non-existent process

tl;dr

As mentioned above, there are two major issues that need to be solved in order to make everyone more productive and turnaround time faster for every PR. I'll start with where we currently are and move to detailing the changes proposed in this PR.

The too many builds issue is a result of the way we are handling the 3 builds required to complete every PR. Each build (aka [status check](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-status-checks)) is enforced using branch policies in strict mode[*](#*---strict-mode). The use of strict mode results in 1) a merge every time there is a change on the target branch and 2) running a new build on each new commit to the branch.

We currently use Mergify to automatically update each branch when there is an update to `master`, or a `release/*`, branch.  This works for the requirements laid out above, in tandem with auto-merge, however has the downside of queuing in the best case scenario[**](#**---best-case-scenario) the summation of `3 * N` where `N = # of Active PRs`. For N=10, that's 165 builds instead of only 30 it would require if we used non-strict mode!

The second major issue is the complication of completing fork-based PRs. The workflow thus far has been to ping me and I will queue the required 3 builds and hope that they don't fail or a different PR merges in before  it.  There is no easy way for anyone else to do this and to be honest it only works for me half the time...

**Summary of changes**

To address each of these issues there are a number of changes we will want to make.

1. Implement a 3-hour policy for all builds. The 3-hour policy will replace the use of "Strict" mode by only invalidating PR status checks if they have not been run within the last 3 hours. The likely-hood of one change breaking another within the 3-hour time-span is slim and will be caught on the target branch if something accidentally makes it through. This is less than ideal as there is a chance changes will make it to `master` that should not but with the time it takes to run all of our CI builds it is currently a necessary evil.

    Implement as a GitHub Action that runs periodically (every 5 minutes) and checks whether any of the builds on the PR have **succeeded** >3 hours ago **and** the PR is out-of-date with the target branch. If those two conditions are true, mark each check as "failed" and leave a message indicating that the build must be re-run.

    With the condition only invalidating if the PR is out-of-date with the target, the easiest way to re-queue the build will be to update the PR with the target branch

    The following change will make the best case scenario reduce to `N*3` builds for `N = # of Active PRs` if they all complete within 3 hours.

1. Update the branch protection policies to only require the "iModel.js Docs" build to run if the only changes in the PR are to the "docs/*" directory.

1. Limit Mergify to only automatically update PR branches if,
    1) PR has all required builds marked with "invalidated after 3 hours"
       - We want to ignore it if there is a failing build for other reasons.
    2) No requested changes
    3) Not a draft

### * - Strict Mode

Strict mode means that the PR branch is required to be up-to-date with the targeted branch (e.g. `master`) before it can be merged.

### ** - Best Case Scenario

Succeed the first time and no merge conflicts

### CODE REVIEW

File: .github/workflows/invalidatePRs.sh [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+  if [[ "open" != $(_jq '.state') ]] || [[ "true" == $(_jq '.draft') ]]; then
+    log "  Skipping due to state=$(_jq '.state') and draft=$(_jq '.draft')."
+    continue
+  fi
+
+  ref=$(_jq '.base.ref')
+  sha=$(_jq '.base.sha')
+
+  if [[ "master" != $ref ]]
```
Comment (calebmshafer) [inline .github/workflows/invalidatePRs.sh:68]: I don't think you need to special case `master` here, right? If you just treat `$ref` always as the base branch name to compare  `$sha`, which is the current position of the PR branch on ref, against the `$refSha` from below that should be enough.

**Review by calebmshafer — COMMENTED** [review]
@saemjeon this looks like a really good start. A couple comments about the workflow for comparing the shas.

Also, I will move the large description I have above into a GitHub issue so that it can be tracked separately from this PR. Once I do that it probably makes sense to create a new PR that is specifically focused on the 3-hour timeout.  Would you mind pushing your changes to a new branch and opening a PR that way?

**Comment by calebmshafer** [comment]
closing in favor of GH issue #1142
